crates/nu-lsp/src/workspace.rs

```diff
@@ -0,0 +1,583 @@
+    pub fn initialize_workspace_folders(&mut self, init_params: Value) -> Result<()> {
+        if let Some(array) = init_params.get("workspaceFolders") {
+            let folders: Vec<WorkspaceFolder> =
+                serde_json::from_value(array.clone()).into_diagnostic()?;
+            for folder in folders {
+                self.workspace_folders.insert(folder.name.clone(), folder);
+            }
+        }
+        Ok(())
+    }
+
+    pub fn rename(&mut self, params: &RenameParams) -> Option<WorkspaceEdit> {
+        let new_name = params.new_name.to_owned();
+        // changes in WorkspaceEdit have mutable key
+        #[allow(clippy::mutable_key_type)]
+        let changes: HashMap<Uri, Vec<TextEdit>> = self
+            .occurrences
+            .iter()
+            .map(|(uri, ranges)| {
+                (
+                    uri.clone(),
+                    ranges
+                        .iter()
+                        .map(|range| TextEdit {
+                            range: *range,
+                            new_text: new_name.clone(),
+                        })
+                        .collect(),
+                )
+            })
+            .collect();
+        Some(WorkspaceEdit {
+            changes: Some(changes),
+            ..Default::default()
+        })
+    }
+
+    /// Goto references response
+    /// TODO: WorkDoneProgress -> PartialResults
+    pub fn references(&mut self, params: &ReferenceParams) -> Option<Vec<Location>> {
+        self.occurrences = BTreeMap::new();
+        let mut engine_state = self.new_engine_state();
+        let path_uri = params.text_document_position.text_document.uri.to_owned();
+        let (mut working_set, id, span, _, _) = self
+            .parse_and_find(
+                &mut engine_state,
+                &path_uri,
+                params.text_document_position.position,
+            )
+            .ok()?;
+        self.find_reference_in_workspace(
+            &mut working_set,
+            &path_uri,
+            id,
+            span,
+            params
+                .work_done_progress_params
+                .work_done_token
+                .to_owned()
+                .unwrap_or(ProgressToken::Number(1)),
+            "Finding references ...",
+        )
+        .ok()?;
+        Some(
+            self.occurrences
+                .iter()
+                .flat_map(|(uri, ranges)| {
+                    ranges.iter().map(|range| Location {
+                        uri: uri.clone(),
+                        range: *range,
+                    })
+                })
+                .collect(),
+        )
+    }
+
+    /// 1. Parse current file to find the content at the cursor that is suitable for a workspace wide renaming
+    /// 2. Parse all nu scripts in the same workspace folder, with the variable/command name in it.
+    /// 3. Store the results in `self.occurrences` for later rename quest
```

Comment:
I love these types of comments. Comments that explain what's supposed to be happening are so valuable to those that have to maintain the code long after the PR.